### PR TITLE
feat(infra): new workspace infra

### PR DESCRIPTION
**Pull request by EYHN**
This PR copying @affine/workspace into common/infra, and adding definitions for services and unit tests.

### CODE REVIEW

**Comment by EYHN** [comment]
Current dependencies on/for this PR:
* `canary`
  * **PR #5562** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/5562?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
    * **PR #5595** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/5595?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
    * **PR #5497** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/5497?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
      * **PR #5563** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/5563?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
        * **PR #5564** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/5564?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
          * **PR #5693** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/5693?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
            * **PR #5617** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/5617?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈
              * **PR #5618** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/5618?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
                * **PR #5565** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/5565?utm_source=stack-comment-icon" target="_blank">Graphite</a> 

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/toeverything/AFFiNE/5617?utm_source=stack-comment).

File: packages/common/infra/src/workspace/manager.ts [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+    return factory.getWorkspaceBlob(metadata.id, blobKey);
+  }
+
+  private instantiate(metadata: WorkspaceMetadata) {
+    logger.info(`open workspace [${metadata.flavour}] ${metadata.id} `);
+    const factory = this.factories.find(x => x.name === metadata.flavour);
+    if (!factory) {
+      throw new Error(`Unknown workspace flavour: ${metadata.flavour}`);
+    }
+    const serviceCollection = this.serviceProvider.collection.clone();
```
Comment (pengx17) [inline packages/common/infra/src/workspace/manager.ts:173]: any advice on when to use `.clone`?

**Review by pengx17 — APPROVED** [review]

File: packages/common/infra/src/workspace/manager.ts (hunk 1 above)
Comment (EYHN) [reply packages/common/infra/src/workspace/manager.ts:173]: `.clone` is quite cheap as it only clones the references.

And since the `ServiceCollection` is only about service definitions, clone it has no side effects.

Clone it here because we will inject some services specific to the current Workspace into the `ServiceCollection` in the `WorkspaceFactory`.

File: .yarn/patches/tinykeys-npm-2.1.0-819feeaed0.patch [hunk 2]
```diff
@@ -0,0 +1,54 @@
+diff --git a/dist/tinykeys.module.js b/dist/tinykeys.module.js
```
Comment (pengx17) [inline .yarn/patches/tinykeys-npm-2.1.0-819feeaed0.patch]: Split the patch in another PR?
